docs/developer/architecture/centralized-cli.rst

```diff
@@ -149,3 +149,203 @@ Cluster Service Administration
+
+Deployment
+""""""""""
+
+Plugin should be developed as one single binary ``kubectl-metalk8s`` available
+from the ISO, easy buildable from GitHub repository and also as a System
+Package for Operating System supported by MetalK8s.
+
+The package should install the plugin in ``/usr/bin`` directory by default.
+
+This plugin should be installed on the bootstrap node by default.
+
+Rejected design choice
+""""""""""""""""""""""
+
+- ``Bash`` kubectl plugin: Bash is great to do simple actions but not when
+  you need to do interaction with some API like Kubernetes API or Salt API.
+- ``Python`` kubectl plugin: Python allows us to do complicated actions and
+  great interactions with APIs but interactions between ``Go`` and Kubernetes
+  are much better, with a sample CLI available.
+
+``metalk8sctl`` CLI
+^^^^^^^^^^^^^^^^^^^
+
+Language
+""""""""
+
+``Python`` is chosen as the language for kubectl plugin for the following
+reasons:
+
+- Ability to interact with Salt directly with Python API.
+- Python installation needed anyway by Salt-minion.
+
+Input and Output
+""""""""""""""""
+
+- Command style:
+
+  .. code-block:: bash
+
+    metalk8sctl <resource> <action>
+
+- Interactive:
+
+  Never.
+
+- Output style:
+
+  Human readable output, do not necessarily need for "machine output" like
+  JSON and YAML.
+
+  Output should extract useful information from Salt output when needed, in
+  case of error only output useful error message(s) from Salt output
+
+.. todo::
+
+  - Define logging information
+
+Salt interaction
+""""""""""""""""
+
+All Salt interaction should be done using
+`Salt Python API <https://docs.saltstack.com/en/latest/ref/clients/index.html>`_
+and not use the ``salt-call``, ``salt``, ``salt-run`` binary at all.
+
+This Python API allows us to interact with Salt-master directly from the host
+machine as Python API directly actis on the Salt socket and does not need to
+execute a command inside the Salt-master container.
+
+Deployment
+""""""""""
+
+``metalk8sctl`` should be available from the ISO and also as a System
+Package for Operating System supported by MetalK8s.
+
+This package should be installed on the bootstrap node by default.
+
+Rejected design choice
+""""""""""""""""""""""
+
+- ``bash`` MetalK8s CLI: Bash is great to do simple actions but not when you
+  need to do interaction with Salt, Salt API, and Kubernetes API.
+- Do not follow kubectl style for the command (``<action> <resource>``), it
+  does not make sense to regroup command per action as actions are really
+  different and this CLI only manages a few resources.
+
+Implementation Details
+----------------------
+
+Two different projects.
+
+First have a simple framework to implement a simple command,
+then each command would update the framework if needed.
+
+TBD !
+
+Iteration 1
+^^^^^^^^^^^
+
+``metalk8sctl``: (everything ?)
+
+- bootstrap deploy
+- bootstrap import-iso ?
+- cluster upgrade ?
+- cluster downgrade ?
+- etcd health
```

Comment:
Honestly, for this one I don't know yet, because we may want to have something that works without salt to check etcd status (using directly the etcd python lib). 
Not sure about this one.

Reply:
I'd say our CLI should focus on Salt implementations, because that's how we'll get the UI to do the same features. Not sure we want to add and `etcd` client in the UI (and even, I don't think that would work, because `etcd` shouldn't be listening to the outside world...).